### PR TITLE
fix: improve confidence score generation

### PR DESCRIPTION
**Pull request by daiwaid**
#### What this PR does / why we need it:
- move the confidence score to be generated after the answer and simplify the prompt
- import prompts from `libem.core.struct.pattern` to `libem.match.prompt` so these prompts are saved as part of benchmark config dump

#### Which issue(s) this PR addresses (if any):

Addresses #

#### Does this PR introduce a user-facing change?

```release-note
```

### CODE REVIEW

**Comment by zenodflow** [comment]
@daiwaid nice! Is there evidence showing the before vs. after results?

E.g., we need some form of calibration evaluation here.

**Comment by daiwaid** [comment]
250 pairs from `amazon-google` using `gpt-4o-2024-08-06`:

Before:
output

After:


Even though the model is in general overconfident in the new results, the wrong answers have lower confidence scores as expected, whereas before, the wrong answers had high confidence scores. In addition, there's about a 15% token saving due to the shortening of the prompt.

**Comment by zenodflow** [comment]
I see. Lgtm

File: libem/match/function.py [hunk 1]
```diff
@@ -312,6 +306,18 @@ def parse_output(output: str) -> dict:
                 confidence = float(''.join(nums))
                 output = output[i + 1:]
                 break
+    
+    for i, line in enumerate(output):
```
Comment (zenodflow) [inline libem/match/function.py:310]: @daiwaid Could we merge this for loop with the for loop above?

File: libem/match/prompt.py [hunk 2]
```diff
@@ -2,6 +2,9 @@
 from libem.core.struct.prompt import (
     Shot, Rules, Experiences
 )
+from libem.core.struct.pattern import (
+    CoT, Confidence
```
Comment (zenodflow) [inline libem/match/prompt.py:6]: Is this used/referenced anywhere?

Comment (zenodflow) [reply libem/match/prompt.py:6]: Ah I see, from the function.py. Good change.

**Comment by daiwaid** [comment]
I revised the parsing logic to account for more outlier cases (when confidence and answer are given on the same line). The code is a bit complicated but this is as simple as I can make it. I included comments so hopefully it's clear.

**Comment by daiwaid** [comment]
The new implementation considers more edge cases that were previously not considered:
1. if the answer/likelihood and the confidence score are given on the same line
2. if the model doesn't give a confidence score, only a likelihood score

Since this parsing logic applies to all models, we can't guarantee that all models will format things the way `gpt-4o` does. Another option is to specify the formatting in the prompt (some kind of guardrail), but that will increase the cost. We can't do structured output since not all models support it.

We can also choose to not deal with these edge cases since they are rare (with gpt models), in that case some slight modifications are still necessary to avoid potentially running into exceptions when calling `output.pop()` when there are no more lines left.

**Comment by zenodflow** [comment]
I see. Thanks. The code path should perhaps split for ones that support structured output (e.g., gpt) and those do not so that the brittleness is contained. Let me read your code again and see if we can simplify / make clearer.

**Comment by zenodflow** [comment]
#109 is now in.

**Review by zenodflow — APPROVED** [review]

**Pull request merged**